torch/csrc/jit/mobile/versioned_operators.cpp

```diff
@@ -0,0 +1,169 @@
+    if (op.has_value()) {
+      fn = [op](Stack& stack) { op->callBoxed(&stack); };
+    } else {
+      return fn;
+    }
+  }
+  return nullptr;
+}
+
+// VersionInfoContainer holds a map of {verion_num: VersionInfo} and an explicit
+// cur_version. cur_version is necessary for the situation where an update
+// invalidate all versions in the container. In such case there's an
+// empty version_vec and the only supported version is cur_version.
+// an example of VersionInfoContainer
+//   version_map =
+//     op_version  | op_info
+//     0           | {"op_A_old", None}
+//     1           | {"op_A", default_val}
+//   cur_version = 2
+struct VersionInfoContainer {
+  VersionInfoContainer(std::unordered_map<int64_t, VersionInfo> version_map, int64_t cur_version)
+      : version_map(version_map), cur_version(cur_version) {}
+
+  // Resolve the operator from the op_version. name is used for messaging.
+  OperatorFunctor resolve_operator(const c10::OperatorName& name, int64_t op_version) const {
+    if (op_version == cur_version) {
+      // It's the current version, no adaptation is needed.
+      return findOperatorFromName(name);
+    }
+    auto it = version_map.find(op_version);
+    if (it != version_map.end()) {
+      // The version is in the supported map, do corresponding adaptation and return
+      const auto& opinfo = it->second;
+      auto fn = findOperatorFromName(opinfo.name);
+      if (opinfo.default_val) {
+        fn = [fn, opinfo](Stack& stack) {
+          stack.push_back(opinfo.default_val.value());
+          fn(stack);
+        };
+      }
+      return fn;
+    }
+    // The version is not supported, throw error with the original name
+    TORCH_CHECK(
+        false,
+        "The version number of ",
+        c10::toString(name),
+        ", ",
+        op_version,
+        " is not compatible in this runtime. ");
+    return nullptr;
+  }
+
+  std::unordered_map<int64_t, VersionInfo> version_map;
+  int64_t cur_version;
+};
+
+// The operator version table that contains the operators with versions that
+// this runtime can "up-version". "up-version" here means that if there's an
+// operator with older version in the model, some treatment can be done to
+// make this model run with current (newer) version of that operator. Currently,
+// this "up-version" is limited to combinations of two situations:
+//   * Route to another op name
+//   * Push additional default values
+// This limitation is guarded by the definition of the VersionInfo struct.
+static const std::unordered_map<std::string, VersionInfoContainer>
+    op_version_table(
+        {
+          {"aten::_convolution",
+            VersionInfoContainer(
+              {{0, VersionInfo(c10::OperatorName("aten::_convolution", ""), true)}}, 1)
+          }
+        });
+} // namespace
+
+OperatorFunctor operator_resolver(
+    const c10::OperatorName& opname,
+    int64_t op_version,
+    int64_t model_version) {
```

Comment:
I see why you needed the model version here. So it's not that we do anything in the new versioning mechanism with the bytecode version; it's just to keep BC with the old way we versioned models, before this new mechanism is implemented.
Basically if we didn't care about about model_version 0x3LL we wouldn't need this parameter at all here.

Reply:
Right. It's for backward compatibility. When there's no model_version 0x3LL we could remove the model version part here.